### PR TITLE
Refactor pipeline executor

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Processors/Executors/ExecutionThreadContext.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+{
+    StoppingPipelineTask * desired = this;
+    StoppingPipelineTask * expected = nullptr;
+
+    while (!data.task.compare_exchange_strong(expected, desired))
+    {
+        if (!expected->executeTask(data, true))
+            return false;
+
+        expected = nullptr;
```
[the remark this answers is not in the document]
Comment (nikitamikhaylov) [reply src/Processors/Executors/ExecutionThreadContext.cpp:24]: Also why we need CAS-loop here if we have mutex?

**Review by nikitamikhaylov — APPROVED** [review]
Ok, but it is hard to read in presence of many `queue.pop()`'. Since we have many kinds of queues like ThreadsQueue, AsyncQueue, why not to introduce another one called Queue. Which will be not just a synonim to std::queue, but will have only three methods: get, push and empty.

**Comment by alexey-milovidov** [comment]
Now performance look alright and tests have passed. Is it still a draft?

**Comment by KochetovNicolai** [comment]
@mergifyio update

**Pull request merged**